Documentation/Evolution/RegexLiteralPitch.md

```diff
@@ -0,0 +1,257 @@
+let regex = {
+  var builder = T.RegexLiteral()
+
+  // __A4 = /([[:alpha:]]\w*)/
+  let __A1 = builder.posixCharacterClassAlpha()
+  let __A2 = builder.characterClassWord()
+  let __A3 = builder.concatenate(__A1, __A2)
+  let __A4 = builder.captureGroup(__A3)
+
+  // __B1 = / = /
+  let __B1 = builder.literal(" = ")
+
+  // __C3 = /([0-9A-F]+)/
+  let __C1 = builder.customCharacterClass(["0"..."9", "A"..."F"])
+  let __C2 = builder.oneOrMore(__C1)
+  let __C3 = builder.captureGroup(__C2)
+
+  let __D1 = builder.concatenate(__A4, __B1, __C3)
+  return T(regexLiteral: builder.finalize(__D1))
+}()
+```
+
+### The `ExpressibleByRegexLiteral` and `RegexLiteralProtocol` protocols
+
+New `ExpressibleByRegexLiteral` and `RegexLiteralProtocol` protocols will be introduced to the standard library, and will serve a similar purpose to the existing literal protocols `ExpressibleByStringInterpolation` and `StringInterpolationProtocol`.
+
+```swift
+public protocol ExpressibleByRegexLiteral {
+  associatedtype RegexLiteral : RegexLiteralProtocol = Regex
+  init(regexLiteral: RegexLiteral)
+}
+
+public protocol RegexLiteralProtocol {
+  init()
+
+  // Informal builder requirements for building a regex literal
+  // will be specified here.
+}
+```
+
+Types conforming to `ExpressibleByRegexLiteral` will be able to provide a custom type that conforms to `RegexLiteralProtocol`, which will be used to build the resulting regex value. A default conforming type will be provided by the standard library (`Regex` here).
+
+Libraries can extend regex handling logic for their domains. For example, a higher-level library could provide linguistically richer regular expressions by incorporate locale, collation, language dictionaries, and fuzzier matching. Similarly, libraries wrapping different regex engines (e.g. `NSRegularExpression`) can support custom regex literals.
+
+### Opting into certain regex features
+
+We intend for the compiler to completely parse [the PCRE syntax][PCRE]. However, types conforming to `RegexLiteralProtocol` might not be able to handle the full feature set. The compiler looks for corresponding function declarations inside `RegexLiteralProtocol` and will emit a compilation error if missing. Conforming types can use `@available` on these function declarations to communicate versioning and add more support in the future.
+
+This approach of lookup combined with availability allows the stdlib to support more features over time.
+
+### Impact of using `/` as the delimiter
+
+#### On comment syntax
+
+Single line comments use the syntax `//`, which would conflict with the spelling for an empty regex literal. As such, an empty regex literal would be forbidden.
+
+Multi-line comments use the `/*` delimiter. As such, a regex literal starting with `*` wouldn't be parsed. This however isn't a major issue as an unqualified `*` is already invalid regex syntax. An escaped `/\*/` regex literal wouldn't be impacted.
+
```

Comment:
A third comment issue that might be worth discussing: If we wanted to support multi-line regex literals, the most natural way to do that given current Swift syntax would be:

```
let regex = ///
            pattern
            ///
```

However, `///` is a comment syntax, and specifically it's the documentation comment syntax. Do we want to rule out this feature, or do we think there's a heuristic we can use to rescue it?

Reply:
(Someone on the overall pitch thread pointed this out—can't remember who, but you may want to give credit.)